### ISSUE TITLE
syntax: Have `prep_init()` output the first-round prepare-message share

### ISSUE THREAD

**Issue by cjpatton** [cfrg/draft-irtf-cfrg-vdaf#60]
Assuming we:
(1) merge https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/53
(2) resolve https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/19 by removing the public parameter

there will be no possibility of a 0-round VDAF. In this case we should consider modifying the VDAF syntax so that `prep_init()` outputs the first prepare-message share. This would resolve some ugliness in the state machine of implementations. For Prio3 for instance, we need to distinguish between a "Ready" state and a "Waiting" state, where the only difference is whether the prepare-message share has been sent. For example, see libprio-rs: https://github.com/divviup/libprio-rs/blob/main/src/vdaf/prio3.rs#L942-L946

@tgeoghegan, @divergentdave, I'd love to have your thoughts here. Note that this change would go in the next draft.

**Comment by tgeoghegan** [comment]
I buy that this would enable 0-round VDAFs (is a 0-round VDAF the same as a DAF?). Can you sketch out how it affects one-or-more round VDAFs? Would the leader still gather up each helper's prepare message share to then join them and broadcast them back out, or would we take the opportunity to have leader send its prepare message to helper in the first message?

**Comment by cjpatton** [comment]
Here's what I had in mind: https://github.com/cfrg/draft-irtf-cfrg-vdaf/commit/ad5ad030bbe6be7a4ecd109686569fb9b2bbb665

Basically the main difference is that `prep_next()` always takes in a prepare message, whereas before in the first round the input would be `None`.

**Comment by bifurcation** [comment]
It seems like the options here are isomorphic.

New->Old: Since the share contents are defined by the concrete protocol, if a protocol couldn't generate a real share in the first instance, it could send an empty value.

Old->New: Similarly, in a zero-round VDAF, the protocol could specify that any putatively non-local inputs to `prep_next()` actually have some value that can be computed locally.  Then an aggregator implementing said protocol could just call `prep_init()` and `prep_next()` together to simulate the proposed modification to `prep_init()`.

Assuming I've got that right, it seems like it doesn't really matter and we should just pick the one that is most natural for the cases we have in hand.

**Comment by cjpatton** [comment]
That's a good way of looking at it, yeah. And I agree they're isomorphic. I think the new syntax would be more natural for te [DAP spec](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-input-share-preparation). In particular it would allow us to rewrite
> ```
> prep_state = VDAF.prep_init(vdaf_verify_param, agg_param, nonce, input_share)
> out = VDAF.prep_next(prep_state, None)
> ```
>
> vdaf_verify_param is the public VDAF parameter, and agg_param is the opaque aggregation parameter. If either step fails, the aggregator marks the report as invalid with error vdaf-prep-error.
>
> Otherwise, the value out is interpreted as follows. If this is the last round of the VDAF, then out is the aggregator's output share. Otherwise, out is the pair (prep_state, prep_msg).

as

> ```
> (prep_state, prep_msg) = VDAF.prep_init(vdaf_verify_key, agg_id, agg_param, nonce, input_share)
> ```
>
> vdaf_verify_key is the secret VDAF verification key shared by both aggregators, agg_id is equal to `0` for the leader and `1` for the helper, and agg_param is the opaque aggregation parameter. If this step fails, the aggregator marks the report as invalid with error vdaf-prep-error.

**Comment by cjpatton** [comment]
I'm not sure why GH decided to close this based on a PR from libprio getting merged. In any case, I've reopened this because we decided not to merge #73 for draft-01. @schoppmp pointed out that this may end up being a regression if we end up having a 0-round VDAF one day. In particular, in order to account for schemes with public keys we will need to either modify the VDAF syntax (basically revert #62) or define a new object ("PK-VDAF" for example). See discussion on #73 for details.
